azure-iot-device/tests/common/test_http_transport.py

```diff
@@ -0,0 +1,279 @@
+        # output = {"response": None, "error": None}
+
+        # def request_callback(response=None, error=None):
+
+        done = transport.request(
+            fake_method, fake_path, mocker.MagicMock(), query_params=fake_query_params
+        )
+        done.result()
+        assert mock_http_client_constructor.call_count == 1
+        assert mock_http_client_request.call_count == 1
+        assert mock_http_client_request.call_args[0][0] == fake_method
+        assert mock_http_client_request.call_args[0][1] == expected_url
+        assert mock_http_client_request.call_args[1]["body"] == ""
+        assert bool(mock_http_client_request.call_args[1]["headers"]) is False
+
+    @pytest.mark.it(
+        "Calls HTTP Client Connection request method with correct method, url, body, and headers."
+    )
+    @pytest.mark.parametrize(
+        "method, path, query_params, body, headers",
+        [
+            pytest.param("__fake_method__", "__fake_path__", None, None, None),
+            pytest.param("__fake_method__", "__fake_path__", "", "", ""),
+            pytest.param("__fake_method__", "__fake_path__", "__fake_query_params__", None, None),
+            pytest.param(
+                "__fake_method__", "__fake_path__", "__fake_query_params__", "__fake_body__", None
+            ),
+            pytest.param(
+                "__fake_method__",
+                "__fake_path__",
+                "__fake_query_params__",
+                "__fake_body__",
+                "__fake_headers__",
+            ),
+        ],
+    )
+    def test_calls_http_client_request_with_given_parameters(
+        self,
+        mocker,
+        transport,
+        mock_http_client_constructor,
+        method,
+        path,
+        query_params,
+        body,
+        headers,
+    ):
+        mock_http_client_request = mock_http_client_constructor.return_value.request
+        if query_params:
+            expected_url = "https://{}/{}?{}".format(fake_hostname, path, query_params)
+        else:
+            expected_url = "https://{}/{}".format(fake_hostname, path)
+
+        cb = mocker.MagicMock()
+        done = transport.request(
+            method, path, cb, body=body, headers=headers, query_params=query_params
+        )
+        done.result()
+        assert mock_http_client_constructor.call_count == 1
+        assert mock_http_client_request.call_count == 1
+        assert mock_http_client_request.call_args[0][0] == method
+        assert mock_http_client_request.call_args[0][1] == expected_url
+
+        actual_body = mock_http_client_request.call_args[1]["body"]
+        actual_headers = mock_http_client_request.call_args[1]["headers"]
+        if body:
+            assert actual_body == body
+        else:
+            assert not bool(actual_body)
+        if headers:
+            assert actual_headers == headers
+        else:
+            assert not bool(actual_headers)
```

Comment:
This could probably be structured better to be less confusing, but it's not the end of the world.